docs: messaging & replay protection (#100)
KILTprotocol/docs: docs/concepts/04_messaging.md

---
id: messaging
title: KILT Messaging
---

Distributed trust on the internet only works if you can exchange credentials and other information securely, and be confident that you are not being fooled or eavesdropped on by bad actors.
To help with that, KILT provides a **transport-agnostic messaging layer** that helps with securely exchanging data between the respective owners of two DIDs.

This messaging layer provides **authenticated end-to-end encryption** – the gold standard in secure communication – so that you do not need to worry about the security of the technologies you use for transporting the message over the internet – be it sending the encrypted message via email, or posting them to and fetching them from a centralized or decentralized messaging service.

To be able to communicate, the two DIDs need to expose **key agreement public keys** for that purpose (a.k.a. an **encryption key**).

> In order to send a message to the other party, a DID owner (let's call her **Alice**) looks up her peer's encryption public key, which can be part of either a [Full-](./02_did.md#full-dids) or a [Light DID](./02_did.md#light-dids).

Reviewer: ```suggestion
In order to send a message to the other party, a DID owner (let's call her **Alice**) looks up her peer's encryption public key, which can be part of either a [Full](./02_did.md#full-dids) or a [Light DID](./02_did.md#light-dids).
```

Using this key in combination with her secret encryption key, **Alice** can now encrypt the message such that only she and the owner of the secret key matching the peer's public key can decrypt it.

The receiving party to whom this message is addressed (say, **Bob**) can decrypt this message after looking up **Alice's** encryption key.
An additional _message authentication code_ (MAC) added during encryption and verified on decryption protects against manipulation of the encrypted data.

> As long as both parties keep their secret keys well protected, the combination of these measures allows **Bob** to be confident that if the message decrypts successfully, it could have only been encrypted by **Alice**, who owns the private key matching the public key advertised on her DID, and has not been read or tampered with by some malicious third party while in transport.

Reviewer: W: I would remove that technical detail in here.

```suggestion
As long as both parties keep their secret keys well protected, the combination of these measures allows **Bob** to be confident that if the message decrypts successfully, it could have only been encrypted by **Alice** and has not been read or tampered with by some malicious third party while in transport.
```

Reviewer: Plus, Bob can be sure that Alice is on the other side if he has had a way to securely receive Alice DID. So we should probably mention this somewhere.

Author: I'm not sure I understand, is this referring to the lookup of the encryption key on chain that must be secure, or that you have to know your peer's DID in the first place to be able to communicate?

> While encrypted, the message travels in a compact and maximally opaque envelope format that only exposes data that the recipient requires to be able to decrypt.

Reviewer: What does **maximally opaque** mean?

Author: reading it again the "maximally" is a bit much. but opaque = not transparent. Could also simply say "privacy preserving" if that's better.

Reviewer 2: I like _privacy preserving_ more. The current version definitely sounds puffy.

Reviewer: But I mean it still reveals information about sender and receiver, which is quite much. How is this more privacy preserving than other encryption schemes?

Reviewer: Different would be, for instance, if the information about the sender is wrapped around an anoncrypt message, where only the information about the receiver is revealed to the public.

Author: nobody's saying it's more privacy-preserving than other encryption schemes - it's privacy-preserving _because_ it's encrypted and exposes only the essentials (and sure, one more layer of wrapping could obscure the sender too, but that is waaaaaaay too much detail for this scope)

Reviewer: Ok then I misunderstood the meaning of the sentence, as this is clearly not a privacy-preserving encryption scheme but, as you said, it preserves some privacy _because_ it is an encryption scheme.

```ts
{
  receiverKeyUri: 'did:kilt:4qMx…66m4#0x4af3…cdcb' // indicates the key to be used when decrypting
  senderKeyUri: 'did:kilt:4qtP…m507#0x65ac…a282' // indicates the key used to encrypt
  ciphertext: '0xdeadbeef…' // message contents in unreadable, encrypted form
  nonce: '0x1234…' // random data required for decryption
}
```

Because we do not simply reference the DIDs of sender and recipient, but the unique identifier of the keys that were used in encryption, this scheme still works if a DID should expose multiple encryption keys from which a message sender may choose.

:::caution

> While no one cannot read or change what is inside an encrypted message even if they intercept it while traveling on the network, a sophisticated attacker may try to guess what is inside and trick you or others by resubmitting a copy of that message later.

Reviewer: ```suggestion
While no one can read nor change what is inside an encrypted message even if they intercept it while traveling on the network, a sophisticated attacker may try to guess what is inside and trick you or others by resubmitting a copy of that message later.
```

> You can learn how to protect against these _replay attacks_ in our [advanced concepts section](./05_Advanced%20Concepts/01_replay_protection.md).

Reviewer: ```suggestion
You can learn how to protect against these _replay attacks_ in our [Advanced Concepts section](./05_Advanced%20Concepts/01_replay_protection.md).
```
